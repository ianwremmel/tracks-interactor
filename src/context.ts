import {Exception} from '@ianwremmel/exception';
import {produce} from 'immer';

/** Thrown when an Interactor is failed via context.fail */
export class InteractorFailure extends Exception {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
  constructor(message: any) {
    super(message);
    if (message instanceof Error) {
      this.original = message;
    }
  }
}

export interface Context<T> {
  data: T;

  error?: InteractorFailure;
}

/** The "arguments" passed to an Interactor */
export class Context<T> {
  /** Indicates if the Interactor has failed */
  get failed(): boolean {
    return !!this.error;
  }

  /** constructor */
  constructor(data: T) {
    this.data = data;
  }

  /**
   * Creates a new Context from the existing. Typically use to produce the
   * return value for Interactor#call.
   *
   * A bit of a vestige from a time when we had Organizers. This method may be
   * removed in favor of returning a plan value.
   */
  extend<R>(fn: (data: T) => R): Context<R> {
    const nextData = produce(this.data, fn);
    const next = new Context(nextData);
    // @ts-expect-error - I'm not sure how to get templating right here, but
    // typechecking works correctly outside of this function, so I think it's
    // fine
    return next;
  }

  /** Marks this Interactor a failed */
  fail(message?: string | Error) {
    if (message instanceof Error) {
      const err = new InteractorFailure(message);
      err.stack = message.stack;
      this.error = err;
      throw err;
    }

    const err = new InteractorFailure(message);
    this.error = err;
    throw err;
  }
}

export interface SuccessContext<T> extends Context<T> {
  failed: false;
}

export interface FailedContext<T> extends Context<T> {
  data: never;
  failed: true;
  error: InteractorFailure;
}

export type ResultContext<T, R> = SuccessContext<R> | FailedContext<T>;
