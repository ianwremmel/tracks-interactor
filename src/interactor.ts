import {Context} from './context';

export interface InteractorConstructor<S, T, R> {
  new (services: S, context: Context<T>): Interactor<S, T, R>;
}

export interface Interactor<S, T, R> {
  services: S;
  context: Context<T>;
  after(): Promise<void>;
  around(fn: () => Promise<void>): Promise<void>;
  before(): Promise<void>;
  rollback?(context: Context<R>): Promise<void>;
}

/** Standardized interface to business logic */
export abstract class Interactor<S, T, R = T> {
  /** constructor */
  constructor(services: S, context: Context<T>) {
    this.services = services;
    this.context = context;
  }
  abstract async call(): Promise<Context<R>>;
}
