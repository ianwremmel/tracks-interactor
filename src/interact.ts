import {Interactor, InteractorConstructor} from './interactor';
import {Context, InteractorFailure} from './context';

/**
 * `interact()` corresponds to the class method `Interactor.call` from ruby, but
 * since TypeScript can't use generic types from a class definition in static
 * methods, we need to do some gymnastics to keep it separate.
 *
 * @param I The interactor or organizer to invoke
 * @param context
 */

export async function interact<S, T, R>(
  services: S,
  I: InteractorConstructor<S, T, R>,
  context: T | Context<T>,
  returnInteractor: boolean
): Promise<{context: Context<R>; interactor: Interactor<S, T, R>}>;

export async function interact<S, T, R>(
  services: S,
  I: InteractorConstructor<S, T, R>,
  context: T | Context<T>
): Promise<Context<R>>;

/** Function for invoking an Interactor (or Organizer) */
export async function interact<S, T, R>(
  services: S,
  I: InteractorConstructor<S, T, R>,
  context: T | Context<T>,
  returnInteractor?: boolean
) {
  const i = new I(
    services,
    context instanceof Context ? context : new Context(context)
  );
  try {
    let ret: Context<R>;

    const inner = async () => {
      if (typeof i.before === 'function') {
        await i.before();
      }
      ret = await i.call();
      if (typeof i.after === 'function') {
        await i.after();
      }
    };

    if (typeof i.around === 'function') {
      await i.around(inner);
    } else {
      await inner();
    }

    // @ts-ignore - typescript thinks ret is being used before it's defined, but
    // it's not.
    if (typeof ret === 'undefined') {
      throw new TypeError(
        'Return context has not been assigned. Did you forget to invoke `fn()` in your `around()` method?'
      );
    }

    if (returnInteractor) {
      return {
        context: ret,
        interactor: i,
      };
    }

    return ret;
  } catch (err) {
    if (err instanceof InteractorFailure) {
      if (returnInteractor) {
        return {
          context,
          interactor: i,
        };
      }
      return err.context;
    }
    throw err;
  }
}
