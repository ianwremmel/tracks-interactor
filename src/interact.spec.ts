import {Context} from './context';
import {interact} from './interact';
import {Interactor} from './interactor';

describe('Kernel', () => {
  describe('Interactor', () => {
    describe('interact()', () => {
      it('turn T into Context<T> when necessary', async () => {
        class A extends Interactor<{}, 1> {
          async call() {
            return this.context;
          }
        }

        const context = await interact({}, A, 1);
        expect(context.data).toEqual(1);
      });

      it('makes context failures throwsafe', async () => {
        const spy = jest.fn();

        class A extends Interactor<{}, any> {
          async call() {
            this.context.fail(new Error('something bad happened'));
            spy();
            return this.context;
          }
        }

        await expect(
          interact({}, A, new Context({}))
        ).resolves.toMatchSnapshot();
        expect(spy).not.toHaveBeenCalled();
      });

      it('invokes an interactor', async () => {
        const spy = jest.fn();

        interface Input {
          a: boolean;
        }

        interface Output {
          b: boolean;
        }

        class A extends Interactor<{}, Input, Output> {
          async call(): Promise<Context<Output>> {
            spy(this.context);
            return this.context.extend(() => ({
              b: true,
            }));
          }
        }

        const out = await interact(
          {},
          A,
          new Context({
            a: false,
          })
        );
        expect(out).toMatchSnapshot();
        expect(spy).toHaveBeenCalled();
        expect(out.data).toEqual({b: true});
      });
    });
  });
});
