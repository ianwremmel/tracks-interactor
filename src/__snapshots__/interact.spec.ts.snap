// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kernel Interactor interact() invokes an interactor 1`] = `
Context {
  "data": Object {
    "b": true,
  },
}
`;

exports[`Kernel Interactor interact() makes context failures throwsafe 1`] = `
Context {
  "data": Object {},
  "error": [InteractorFailure: Wrapped Exception: something bad happened],
}
`;
